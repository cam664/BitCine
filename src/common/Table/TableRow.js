import React from 'react';

const TableRow = ({ rowData, columns }) => {
  return (
    <tr>
      {columns.map(col =>
        <td>
          {rowData[col.field]}
        </td>  
      )}
    </tr>
  );
}

export default TableRow;