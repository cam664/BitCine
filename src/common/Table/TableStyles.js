export default {
  tableMainWrap: `
    table {
      width: 100%;
      border-collapse: collapse;
      tr {
        width: 100%;
      }
      th, td {
        div, a {
          padding: 15px 7.5px;
        }
      }
      tr td, tr th {
        &:first-of-type div,
        &:first-of-type a {
          padding-left: 15px;
        }
        &:last-of-type div,
        &:last-of-type a {
          padding-right: 15px;
        }
      }
      th {
        text-align: left;
      }
    }
  `,
  tableBodyWrap: `
    max-height: 450px;
    overflow: hidden scroll;
    tr {
      background: #fff;
      border-bottom: 1px solid #f2f3f3;
    }
  `,
  rowPlaceholder: `
    display: inline-block;
    width: 25%;
    height: 100%;
    border-radius: 6px;
    min-width: 30px;
    background-image: linear-gradient(to top, #f2f3f3, #f6f6f6);
  `,
  isClickable: `
    a {
      display: block;
      text-decoration: none;
      color: #000;
    }
    &:hover {
      cursor: pointer;
      background-color: rgba(187,221,255,0.3);
    }
  `
}