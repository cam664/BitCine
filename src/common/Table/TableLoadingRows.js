import React from 'react';
import TableRow from './TableRow';

export const TableLoadingRows = ({ columns, numOfRows }) => {
  
  return (
    <>
      {Array(numOfRows).fill(
        <TableRow arePlaceholders columns={columns} />
      )}
    </>
  );
}