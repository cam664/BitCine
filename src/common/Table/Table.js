import React, { useRef } from 'react';
import TableRow from './TableRow';

/** @jsx jsx */
import { css, jsx } from '@emotion/core';
import styles from './TableStyles';
import { TableLoadingRows } from './TableLoadingRows';

const Table = ({ data, columns, isLoading, onMore }) => {
  const scrollWrapRef = useRef();

  const handleScroll = (e) => {
    const scrollPos = e.currentTarget.scrollTop;
    const scrollEnd = (scrollWrapRef.current.scrollHeight - scrollWrapRef.current.clientHeight);
    if (scrollPos >= scrollEnd) {
      onMore();
    }
  }

  return (
    <section css={css`${styles.tableMainWrap}`}>
      <table>
        <thead>
          <tr>
            {columns.map(col => 
              <th key={col.label} style={{width: col.width}}><div>{col.label}</div></th>
            )}
          </tr>
        </thead>
      </table>
      
      <div onScroll={handleScroll} ref={scrollWrapRef} css={css`${styles.tableBodyWrap}`}>
        <table>
          <tbody>
            {data.map((rowData, i) => 
              <TableRow key={i} isClickable={rowData.model ? false : true} rowData={rowData} columns={columns} />
            )}
            {isLoading && 
              <>
                <TableLoadingRows columns={columns} numOfRows={3} />
              </>
            }
          </tbody>
        </table>
      </div>
    </section>
  );
}

export default Table;