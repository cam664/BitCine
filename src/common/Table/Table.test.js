import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Table from './Table';
import { BrowserRouter as Router  } from 'react-router-dom';

const columns = [
  {
    label: 'Column 1',
    field: 'model',
    width: '40%'
  },
  {
    label: 'Column 2',
    field: 'name',
    width: '20%'
  }
];
const data = [
  {
    model: 'Data Placeholder',
    name: 'Data Placeholder'
  },
  {
    model: 'Data Placeholder',
    name: 'Data Placeholder'
  }
]

describe('Table', () => {
  it('renders a snapshot', () => {
    const component = renderer.create(
      <Router>
        <Table columns={columns} data={data} />
      </Router>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('header columns', () => {
    it('should render a table header cell for each column', () => {
      const component = mount(
        <Router>
          <Table columns={columns} data={data} />
        </Router>
      )

      expect(
        component.find('th')
      ).toHaveLength(columns.length)
    })

    it('should use the label property as a title', () => {
      const columns = [
        {
          label: 'Column Label',
          field: 'model'
        }
      ]
      const data = [
        {
          model: 'First row',
        },
        {
          model: 'Second row',
        }
      ]
      const component = mount(
        <Router>
          <Table columns={columns} data={data} />
        </Router>
      )
  
      const cell = component.find('th')
  
      expect(cell.text()).toEqual('Column Label')
    })
  });

  it('should update the table when data is added', () => {
    const columns = [
      {
        label: 'Column Label',
        field: 'model'
      }
    ]
    const data = [
      {
        model: 'First row',
      },
      {
        model: 'Second row',
      }
    ]

    const component = mount(
      React.createElement(
        props => (
          <Router>
            <Table {...props} />
          </Router>
        ),
        { columns, data }
      )
    )

    const cellsText = component
      .find('tbody td')
      .map(element => element.text())

    const dataText = data.map(item => item.model)

    expect(cellsText).toEqual(dataText)

    data.push({
      model: 'Third row'
    })

    component.setProps({ data })

    const updatedCellsText = component
      .find('tbody td')
      .map(element => element.text())

    const updatedDataText = data.map(item => item.model)

    expect(updatedCellsText).toEqual(updatedDataText)
  })

});

