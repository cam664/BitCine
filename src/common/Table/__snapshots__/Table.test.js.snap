// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders a snapshot 1`] = `
<section
  className="css-1rt94nh-Table"
>
  <table>
    <thead>
      <tr>
        <th
          style={
            Object {
              "width": "40%",
            }
          }
        >
          <div>
            Column 1
          </div>
        </th>
        <th
          style={
            Object {
              "width": "20%",
            }
          }
        >
          <div>
            Column 2
          </div>
        </th>
      </tr>
    </thead>
  </table>
  <div
    className="css-1mzwwa2-Table"
    onScroll={[Function]}
  >
    <table>
      <tbody>
        <tr>
          <td
            style={
              Object {
                "width": "40%",
              }
            }
          >
            <div>
              Data Placeholder
            </div>
          </td>
          <td
            style={
              Object {
                "width": "20%",
              }
            }
          >
            <div>
              Data Placeholder
            </div>
          </td>
        </tr>
        <tr>
          <td
            style={
              Object {
                "width": "40%",
              }
            }
          >
            <div>
              Data Placeholder
            </div>
          </td>
          <td
            style={
              Object {
                "width": "20%",
              }
            }
          >
            <div>
              Data Placeholder
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;
